tree/treeplayer/src/TCsvDS.cxx

```diff
@@ -0,0 +1,255 @@
+      type = "bool";
+   } else { // everything else is a string
+      type = "std::string";
+   }
+   // TODO: Date
+
+   fColTypes[fHeaders[idxCol]] = type;
+}
+
+std::vector<std::string> TCsvDS::ParseColumns(std::string &line)
+{
+   std::vector<std::string> columns;
+
+   for (size_t i = 0; i < line.size(); ++i) {
+      i = ParseValue(line, columns, i);
+   }
+
+   return columns;
+}
+
+size_t TCsvDS::ParseValue(std::string &line, std::vector<std::string> &columns, size_t i)
+{
+   std::stringstream val;
+   bool quoted = false;
+
+   for (; i < line.size(); ++i) {
+      if (line[i] == fDelimiter && !quoted) {
+         break;
+      } else if (line[i] == '"') {
+         // Keep just one quote for escaped quotes, none for the normal quotes
+         if (line[i + 1] != '"') {
+            quoted = !quoted;
+         } else {
+            val << line[++i];
+         }
+      } else {
+         val << line[i];
+      }
+   }
+
+   columns.emplace_back(val.str());
+
+   return i;
+}
+
+TCsvDS::TCsvDS(std::string_view fileName, bool readHeaders, char delimiter) // TODO: Let users specify types?
+   : fFileName(fileName),
+     fDelimiter(delimiter)
+{
+   std::ifstream stream(fFileName);
+   std::string line;
+
+   // Read the headers if present
+   if (readHeaders) {
+      if (std::getline(stream, line)) {
+         FillHeaders(line);
+      } else {
+         auto msg = "Error reading headers of CSV file " + fileName;
+         throw std::runtime_error(msg);
+      }
+   }
+
+   if (std::getline(stream, line)) {
+      auto columns = ParseColumns(line);
+
+      // Generate headers if not present
+      if (!readHeaders) {
+         GenerateHeaders(columns.size());
+      }
+
+      // Infer types of columns with first record
+      InferColTypes(columns);
+
+      // Read all records and store them in memory
+      do {
+         fRecords.emplace_back();
+         FillRecord(line, fRecords.back());
+      } while (std::getline(stream, line));
+   }
+}
+
+TCsvDS::~TCsvDS()
+{
+   for (auto &record : fRecords) {
+      for (size_t i = 0; i < record.size(); ++i) {
+         void *p = record[i];
+         auto &colType = fColTypes[fHeaders[i]];
+
+         if (colType == "int") {
+            delete static_cast<int *>(p);
+         } else if (colType == "double") {
+            delete static_cast<double *>(p);
+         } else if (colType == "bool") {
+            delete static_cast<bool *>(p);
+         } else {
+            delete static_cast<std::string *>(p);
+         }
+      }
+   }
+}
+
+const std::vector<std::string> &TCsvDS::GetColumnNames() const
+{
+   return fHeaders;
+}
+
+const std::vector<std::pair<ULong64_t, ULong64_t>> &TCsvDS::GetEntryRanges() const
+{
+   return fEntryRanges;
+}
+
+std::string TCsvDS::GetTypeName(std::string_view colName) const
+{
+   if (!HasColumn(colName)) {
+      auto e = "The dataset does not have column " + colName;
+      throw std::runtime_error(e);
+   }
+
+   return fColTypes.at(colName.data());
```

Comment:
We should use operator[] for performance reasons rather than at. We are sure we are not out of bounds after all.

Reply:
Turns out the only way to access the map with `operator[]` here is to construct a string from the string_view and pass it as the key. This is potentially more costly than the boundary check of `at`, so after discussing with @bluehood I will leave it as it is, since `GetTypeName` is just invoked once per column anyway.